app/src/debugger.lisp

```diff
@@ -0,0 +1,304 @@
+     (when *display*
+       (show-amplitudes)))
+    (t
+     (error "No program has been loaded."))))
+
+(defun program-finished-p ()
+  (or (null (qvm::pc *qvm*))
+      (>= (qvm::pc *qvm*)
+          (qvm::loaded-program-length *qvm*))))
+
+(define-command step ()
+  "Usage: step
+
+Run the next instruction and stop."
+  (error-unless-qvm-exists)
+  (cond
+    ((program-finished-p)
+     (format t "Finished program execution.~%"))
+    (t
+     (format t "~/quil:instruction-fmt/~%" (qvm::current-instruction *qvm*))
+     (setf *qvm* (qvm:transition *qvm* (qvm::current-instruction *qvm*)))
+     (when *display*
+       (print-amplitudes)))))
+
+(define-command continue ()
+  "Usage: continue
+
+Resume program execution from the current instruction."
+  (error-unless-qvm-exists)
+  (cond
+    ((program-finished-p)
+     (format t "Finished program execution.~%"))
+    (t
+     (loop :for pc := (qvm::pc *qvm*)
+           :for counter :from 0
+           :for breakpoint-p := (and (find pc *breakpoints*)
+                                     (plusp counter))
+           :until (or (program-finished-p) breakpoint-p) :do
+             (setf *qvm* (qvm:transition *qvm* (qvm::current-instruction *qvm*)))
+           :finally (when breakpoint-p
+                      (format t "Stopping at breakpoint in instruction ~D:~%~/quil:instruction-fmt/~%"
+                              pc (qvm::current-instruction *qvm*)))
+                    (return *qvm*)))))
+
+(define-command load (filename)
+  "Usage: load FILENAME
+
+Load a program instantiating a suitable QVM."
+  (unless filename
+    (error "File name not specified."))
+  (let* ((program (quil:read-quil-file (string-trim " " filename)))
+         (number-of-qubits (quil:qubits-needed program)))
+    (format t "Read ~A using ~D qubits.~%" filename number-of-qubits)
+    (setf *source-filename* filename
+          *qvm* (qvm:make-qvm number-of-qubits
+                              :classical-memory-model (qvm:memory-descriptors-to-qvm-memory-model
+                                                       (quil:parsed-program-memory-definitions program)))
+          *breakpoints* nil)
+    (qvm:load-program *qvm* program))
+  (setf *breakpoints* nil))
+
+(define-command reload ()
+  "Usage: reload
+
+Re-read the file corresponding to the previously loaded program."
+  (unless *source-filename*
+    (error "No program has been loaded."))
+  (format t "Reloading ~S.~%" *source-filename*)
+  (%load *source-filename*))
+
+(define-command break (instruction-index)
+  "Usage: break INSTRUCTION-INDEX
+
+Set up a breakpoint at the specified instruction."
+  (error-unless-qvm-exists)
+  (let ((idx (parse-integer instruction-index)))
+    (unless (and (integerp idx)
+                 (not (minusp idx))
+                 (< idx (length (qvm::program *qvm*))))
+      (error "Invalid instruction index ~S." idx))
+    (pushnew idx *breakpoints*)
+    (setf *breakpoints* (sort *breakpoints* #'<))))
```

Comment:
Why do you want this to be sorted?

Reply:
You're right. I don't. Thanks for pointing that out.